lib/services/local/configuration.go

```diff
@@ -221,24 +225,89 @@ func (s *ClusterConfigurationService) DeleteClusterConfig() error {
 }
 
 // SetClusterConfig sets services.ClusterConfig on the backend.
-func (s *ClusterConfigurationService) SetClusterConfig(c services.ClusterConfig) error {
-	value, err := services.MarshalClusterConfig(c)
+func (s *ClusterConfigurationService) SetClusterConfig(newConfig services.ClusterConfig) error {
+	ctx := context.TODO()
+
+	storedConfig, err := s.GetClusterConfig()
+	if err != nil && !trace.IsNotFound(err) {
+		return trace.Wrap(err)
+	}
+
+	// Reapply the stored override to the new config unless it already has
+	// its own overide specified.
+	if storedConfig != nil && storedConfig.HasOverride() && !newConfig.HasOverride() {
+		storedOverride, err := storedConfig.ExtractOverride()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if err := newConfig.ApplyOverride(storedOverride); err != nil {
+			s.log.WithError(err).Warnf("Failed to fully apply override %v to %v.", storedOverride, newConfig)
```

Comment:
I think this should be an error returned to the user. Otherwise may result in the surprise behavior for users due to silent side-effect.

Reply:
Note that this is `SetClusterConfig` that is also called from the auth server initialization procedure.

I didn't make this into a proper error to guarantee the more authoritative config-file `ClusterConfig` is accepted even if there already is an override that cannot be (fully) applied to the new `ClusterConfig`. For example, if there was a `session_recording` override (applicable only if the config-file `session_recording` is one of `off, node, node-sync`) and an administrator changed the config-file setting to `proxy`, the auth server would refuse to initialize upon next start up.

The solution presented here is to always apply the more authoritative `ClusterConfig`, dropping those parts of the override that aren't effective anymore, and to log a warning about the situation.

An alternative would be to properly fail as per your suggestion, and force the administrator to manually change/remove the `ClusterConfigOverride` resource so that he can introduce his settings. I wouldn't consider that approach appropriate though.